### ISSUE TITLE
nx doesn't build the deps projects when they change

### ISSUE THREAD

**Issue by wizardnet972** [nrwl/nx#15113]
### Current Behavior

I build a repo `nx-swc`  with `cli` and `utils` swc libs (both buildable).

When I change [this file](https://github.com/wizardnet972/nx-swc/blob/main/packages/utils/src/lib/utils.ts#L2) in `utils` project `cli` doesn't affect by this change.

In the cli I add [`serve` executor target](https://github.com/wizardnet972/nx-swc/blob/main/packages/cli/project.json#L8) (`@nrwl/js:node`) and I run it `yarn nx serve cli`, when I change the file in utils the cli not build again.

The `cli` is depended on utils [by this import](https://github.com/wizardnet972/nx-swc/blob/main/packages/cli/src/lib/cli.ts#L1) I would expect from nx to know that dependency has changed.

### Expected Behavior

When repo dependency change, should rebuild the project.

### GitHub Repo

https://github.com/wizardnet972/nx-swc

### Steps to Reproduce

1. clone the project.
2. run yarn install
3. run yarn nx serve cli
4. make some change in `packages/utils/src/lib/utils.ts` and then save the file.
 you can see the cli doesn't rebuild.

-- another scenario --
1. run yarn nx run-many --target build --all
2. run yarn nx serve cli
3. check for the console
4. change the output of the console in `packages/utils/src/lib/utils.ts` and then save the file.
5. run yarn nx serve cli again and you can see that nothing change.

### Nx Report

```shell
>  NX   Report complete - copy this into the issue template

   Node : 18.13.0
   OS   : darwin arm64
   npm  : 9.3.0

   nx                      : 15.7.2
   @nrwl/jest              : 15.7.2
   @nrwl/linter            : 15.7.2
   @nrwl/workspace         : 15.7.2
   @nrwl/cli               : 15.7.2
   @nrwl/devkit            : 15.7.2
   @nrwl/eslint-plugin-nx  : 15.7.2
   @nrwl/js                : 15.7.2
   @nrwl/tao               : 15.7.2
   typescript              : 4.8.4

✨  Done in 0.97s.
```
```


### Failure Logs

_No response_

### Additional Information

_No response_

**Comment by owl-app** [comment]
@wizardnet972 the same problem

**Comment by pawelkeska** [comment]
@wizardnet972 and when i remove package front node_modules everything works

**Comment by wizardnet972** [comment]
Hi @pawelkeska, please explain what you mean. 
The problem with this issue is I create two packages (in the project): cli and utils. When I changed the utils, I was expecting to cli to re-build, but not it didn't.

**Comment by jon9090** [comment]
I am encountering the same issue where Nx fails to rebuild the package even when its library dependencies have changed. 

This poses a challenge for utilizing Nx with interconnected multiple packages. 
This problem is of utmost importance for anyone utilizing Nx.

**Comment by SebasG22** [comment]
I'm getting the same error on the latest version. It's weird because this was working in previous versions and now the most simple example does not work.

Also, this can be related to #16030. 

Any help would be greatly appreciated.
@FrozenPandaz  @AgentEnder  @mandarini

**Comment by jon9090** [comment]
@SebasG22 you can vote in the @wizardnet972  post to promote the issue

**Comment by wizardnet972** [comment]
@SebasG22, I had reported the same issue. 

The problem occurs when you create a ts preset app and library. 
Upon running the app for the first time, you encounter the "Module not found" error as the library was not yet built. 
Once you build the library and run the app again, the app doesn't rebuild upon changes as Nx fails to detect the changes in the library, which is problematic.

I agree with @jon9090. How developers who need to integrate multiple libraries can work with this preset? Typically, I split the application or library logic into smaller libs, such as cli and utils, to address this issue.

This problem has prevented me from using Nx for creating and publishing libraries, and I'm uncertain why it's not given higher priority.

**Comment by jaysoo** [comment]
@ikosta Can you open a new issue to track that, with a reproduction repo? It's a separate issue.

It looks like file change events are not working for dependencies, which is why the process is not reloaded. We'll look into this.

**Comment by jaysoo** [comment]
Took a closer look, and the problem is two-fold:

1. File change events are not watching for changes in dependencies (as noted above).
2. When running re-compiling the parent project (`cli` in the example), dependencies are not re-compiled.

(1) is a relatively simple fix if we ask the Nx daemon to also notify on dependency changes. (2) needs some work to make sure the dependency is also built during watch.

**Comment by wizardnet972** [comment]
@jaysoo, I greatly appreciate your efforts in looking into this matter. I was wondering if the issue I reported at https://github.com/nrwl/nx/issues/16530 is connected to the current problem we are facing. My suspicion stems from the way nx handles dependency resolution.

**Comment by jaysoo** [comment]
@wizardnet972 Thats' a different issue. This issue is about the watch mode not rebuilding dependencies and re-launching the process. I'll take a look at the other issue.

**Comment by wizardnet972** [comment]
@jaysoo Could you please suggest a possible solution for this issue? Additionally, I would appreciate it if you could provide an update on when the resolution is expected

**Comment by alxrodav** [comment]
Any update here?

**Comment by AshHeskes** [comment]
I'm seeing this too. It seems the logical order of how commands are ran is backwards from expectation..

e.g. for a project like:

```txt
+ app1
+ lib1
+ lib2
```

Where:

+ `app1` depends on  `lib2` and `lib1`
+ `lib2` depends on `lib1`.

Then starting e.g. `nx watch --all -- nx run \$NX_PROJECT_NAME:build`:

+ If you change `lib1` only `lib1` gets rebuilt.
+ If you then change `app1`  then `lib2` gets rebuilt, then `app1` gets rebuilt.

You expect it to work the other way. Where:

+ If you change `lib1` then `lib1` gets rebuilt, then `lib2` gets rebuilt, then `app1` gets rebuilt.
+ if you then change `app1` then only `app1` gets rebuilt as nothing depends on it.

In other words the current behaviour seems to run commands on a fan in basis, where people expect it to fan out.

**Comment by bobalazek** [comment]
I have started experiencing the same issue. For me it seem to have worked fine in v16.5.0, but when I upgraded to v16.7.2 it stopped working. Any solutions for that, besides a downgrade?

**Comment by biz812** [comment]
We have also begun to experience this issue, like the previous post we went from v16.5.4 to v16.7.4.
This issue has been open since well before the release of 16.5.x but we haven't experienced it until now... I wonder if this is a regression somewhere?

**Comment by jonlambert** [comment]
Apologies for the ping too, but this is a huge issue to encounter after dedicating time to moving to Nx. I'm noticing this specifically in a release process – when a dependency is updated, I would hope that the dependant would have the new version referenced and be rebuilt/released too?

I thought this was something we were doing wrong with configuration. Is there any acknowledgement of this issue?

**Comment by jellebrouwer** [comment]
In the options of the serve target: "runBuildTargetDependencies": true

**Comment by noelsoong** [comment]
updated from 18.3.3 to 19.6.3 and got this issue.

**Comment by yossiyedid** [comment]
same here, any updates?

**Comment by net** [comment]
Turborepo doesn't have this issue. I'd like to use Nx, but this bug is a blocker to doing so.

The fact that this has been an issue for over a year suggests that the level of maintenance Nx gets is not suitable for production.

**Comment by net** [comment]
As far as I'm aware nothing has been done to fix this. Would love to find out I'm wrong.

**Comment by monabbous** [comment]
I have resolved this

# tldr; the fix:
 
create a file `nx.js` in your root of your workspace
```js
// DO NOT EDIT THIS FILE !!
// this file is to correct the error when running serve target of "@nx/js:node" where
// it doesn't run the build executor 

const { execSync } = require('child_process');

execSync(`nx ${process.argv.slice(2).join(' ')}`, {
  stdio: 'inherit',
  env: {
    ...process.env,
    NX_CLI_SET: 'true',
  },
});

```

note that I have nx command installed globally, you may need to do so or replace nx with how you run your targets

# Explanation, please fix this nx team

The problem is to do with the way that js:node is running the build executor.
within the code that is running the build executor, it is calling the nx command as such

```
fork([require.resolve('nx'), '<build target>', '<build target options>'])
```

`require.resolve('nx')` is expected to resolve the binary file `nx.js` to execute the command, but since now new projects has the `nx.json` file in the root level of the workspace, due to how the precedence order of the resolve function work it would resolve this command as:
```
node nx.json <build target> <build target options>
```

I'm not sure what is the best way to resolve this by the team, but maybe by using something other than the resolve.require.

I hope this helps.

chears
